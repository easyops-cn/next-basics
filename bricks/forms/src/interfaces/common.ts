export interface Option {
  label: string;
  caption?: string;
  value: any;
}

export interface OptionGroup {
  label: string;
  options: Option[];
}

export type OptionType = Option | OptionGroup;
