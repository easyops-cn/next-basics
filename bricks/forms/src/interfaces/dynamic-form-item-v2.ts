import { UseBrickConf } from "@next-core/brick-types";
import { GeneralComplexOption } from "@next-libs/forms";
import {
  CascaderExpandTrigger,
  CascaderOptionType,
  FieldNamesType,
} from "antd/lib/cascader";
import { Rule, RuleObject } from "antd/lib/form";
import { StoreValue } from "antd/lib/form/interface";
import { OptionType } from ".";

export enum ComponentType {
  INPUT = "input",
  INPUT_NUMBER = "inputNumber",
  INPUT_PASSWORD = "inputPassword",
  SELECT = "select",
  CASCADER = "cascader",
}

interface BasicProps {
  /**
   * 是否禁用状态
   */
  disabled?: boolean | ((row: Record<string, any>, index: number) => boolean);
  placeholder?: string;
}

export interface InputProps extends BasicProps {
  /**
   * 声明 input 类型，同原生 input 标签的 type 属性
   */
  type?: string;
  /**
   * 最大长度
   */
  maxLength?: number;
  /**
   * 可以点击清除图标删除内容
   */
  allowClear?: boolean;
}

export interface InputNumberProps extends BasicProps {
  /**
   * 最大值
   */
  max?: number;
  /**
   * 最小值
   */
  min?: number;
  /**
   * 每次改变步数，可以为小数
   */
  step?: number | string;
  /**
   * 数值精度
   */
  precision?: number;
}

export interface InputPasswordProps extends BasicProps {
  /**
   * 是否显示切换按钮
   */
  visibilityToggle?: boolean;
}

export interface SelectProps extends BasicProps {
  /**
   * 可以点击清除图标删除内容
   */
  allowClear?: boolean;
  /**
   * 设置 Select 的模式为多选或标签
   */
  mode?: "multiple" | "tags";
  /**
   * 候选项列表
   */
  options:
    | GeneralComplexOption<string | number>[]
    | GeneralComplexOption<string | number>[][];
  /**
   * 支持搜索
   */
  showSearch?: boolean;
  /**
   * 基于 `options` 列表中的某个字段进行分组显示
   */
  groupBy?: string;
  /**
   * 在 mode 为 `tags` 和 `multiple` 模式下自动分词的分隔符
   */
  tokenSeparators?: string[];
  /**
   * 最多显示多少个 tag
   */
  maxTagCount?: number | "responsive";
  /**
   * 下拉选项的渲染方式
   */
  popoverPositionType?: "default" | "parent";
  /**
   * 支持在文本后添加自定义构件 具体查看 [UseBrickConf](/next-docs/docs/api-reference/brick-types.usesinglebrickconf
   */
  suffix?: UseBrickConf;
  /**
   * 设置后置构件容器的样式
   */
  suffixStyle?: React.CSSProperties;
}

export interface CascaderProps extends BasicProps {
  /**
   * 是否支持清除
   */
  allowClear?: boolean;
  /**
   * 候选项
   */
  options: CascaderOptionType[] | CascaderOptionType[][];
  /**
   * 次级菜单的展开方式，可选 `click` 和 `hover`
   */
  expandTrigger?: CascaderExpandTrigger;
  /**
   * 浮层预设位置
   */
  popupPlacement?: "bottomLeft" | "bottomRight" | "topLeft" | "topRight";
  /**
   * 支持搜索
   */
  showSearch?: boolean;
  /**
   * 自定义 options 中 label value children 的字段
   */
  fieldNames?: FieldNamesType;
}

export interface EditorProps {
  /**
   * 占位符
   */
  placeholder?: string;
  /**
   * 语言模式
   */
  mode?: string;
  /**
   * 最小行数
   */
  minLines?: number;
  /**
   * 最大行数
   */
  maxLines?: number;
  /**
   * 主题
   */
  theme?: string;

  /**
   * 显示左侧区域槽
   */
  showGutter?: boolean;
  /**
   * 只读模式
   */
  readOnly?: boolean;
  /**
   * 显示行数
   */
  showLineNumbers?: boolean;
  /**
   * 显示复制按钮
   */
  showCopyButton?: boolean;
  /**
   * 显示展开按钮
   */
  showExpandButton?: boolean;
  /**
   * 是否开启自动补全
   */
  enableLiveAutocompletion?: boolean;
  /**
   * 显示打印边距
   */
  printMargin?: boolean;
  /**
   * 自定义自动补全
   */
  customCompleters?: string[];
}

export interface AutoCompleteProps extends BasicProps {
  /**
   * 是否支持清除
   */
  allowClear?: boolean;
  /**
   * 候选项
   */
  options: string[] | OptionType[];
}

export interface InputColumn {
  // type: ComponentType.INPUT;
  type: "input";
  props: InputProps;
}

export interface SelectColumn {
  // type: ComponentType.SELECT;
  type: "select";
  props: SelectProps;
}

export interface InputNumberColumn {
  // type: ComponentType.INPUT_NUMBER;
  type: "inputNumber";
  props: InputNumberProps;
}

export interface InputPasswordColumn {
  // type: ComponentType.INPUT_PASSWORD;
  type: "inputPassword";
  props: InputPasswordProps;
}

export interface CascaderColumn {
  // type: ComponentType.CASCADER;
  type: "cascader";
  props: CascaderProps;
}

export interface EditorColumn {
  type: "editor";
  props: EditorProps;
}

export interface AutoCompleteColumn {
  type: "autoComplete";
  props: AutoCompleteProps;
}

export interface BasicColumn {
  /**
   * 表单项字段说明
   */
  label?: string;
  /**
   * 表单项字段名
   */
  name: string;
  /**
   * 表单项校验规则
   */
  rules?: ({ unique: boolean; message: string } & Rule & {
      message?: string;
      validator: (
        rule: RuleObject,
        value: StoreValue,
        callback: (error?: string) => void,
        fullValue: {
          formValue: Record<string, any>[];
          rowValue: Record<string, any>;
          rowIndex: number;
        }
      ) => Promise<void | any> | void;
    })[];
  /**
   * 表单项所占份额
   */
  flex?: string | number;
  /**
   * 表单项默认值
   */
  defaultValue?: any;
}

export type Column = BasicColumn &
  (
    | InputColumn
    | InputNumberColumn
    | InputPasswordColumn
    | SelectColumn
    | CascaderColumn
    | EditorColumn
    | AutoCompleteColumn
  );
