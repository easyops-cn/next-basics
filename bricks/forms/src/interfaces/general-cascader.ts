import { CascaderOptionType } from "antd/lib/cascader";

export interface ProcessedOptionData {
  layerIndex: number;
  curOption: CascaderOptionType;
  selectedOptions: CascaderOptionType[];
}
