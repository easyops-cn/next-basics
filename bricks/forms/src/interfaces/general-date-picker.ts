export interface DisabledDate {
  second?: string | number;
  minute?: string | number;
  hour?: string | number;
  date?: string | number;
  month?: string | number;
  year?: string | number;
  weekday?: string | number;
}

export type DisabledDateType = DisabledDate | DisabledDate[];
