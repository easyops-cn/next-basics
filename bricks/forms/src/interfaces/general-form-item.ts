import { UseBrickConf } from "@next-core/brick-types";

export interface ControlBrickConfig {
  /**作为表单控件展示的构件配置 */
  useBrick: UseBrickConf;
}
