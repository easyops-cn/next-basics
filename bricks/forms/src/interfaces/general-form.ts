export enum FormAlignment {
  Left = "left",
  Right = "right",
  Center = "center",
}

export type UiType = "default" | "dashboard";
