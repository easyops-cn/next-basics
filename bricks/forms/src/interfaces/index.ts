export * from "./general-form";
export * from "./common";
export * from "./upload-files-v2";
export * from "./general-date-picker";
export * from "./general-cascader";
export * from "./dynamic-form-item-v2";
