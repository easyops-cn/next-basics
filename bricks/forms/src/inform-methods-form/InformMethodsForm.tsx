/* eslint-disable react/display-name */

import React, { forwardRef, useState, useEffect } from "react";
import { Checkbox } from "antd";

import { FormItemWrapperProps, FormItemWrapper } from "@next-libs/forms";
import { CustomSenderApi_listSupportInform } from "@next-sdk/msgsender-sdk";

import { InformMethod } from "../interfaces";

export interface InformMethodsFormItemProps {
  value?: string[];
  children?: React.ReactNode;
  informMethodList?: InformMethod[];
  onChange?: (value: any) => void;
}

export const InformMethodsFormItem = forwardRef<
  HTMLDivElement,
  InformMethodsFormItemProps
>((props, ref) => {
  const { value, informMethodList, onChange } = props;

  const renderGroup = (): React.ReactElement => {
    const options = informMethodList.map((informMethods) => ({
      label: informMethods.description,
      value: informMethods.inform_type,
    }));

    return (
      <Checkbox.Group options={options} onChange={onChange} value={value} />
    );
  };

  return <span ref={ref}>{informMethodList && renderGroup()}</span>;
});

export interface InformMethodsFormProps extends FormItemWrapperProps {}

export function InformMethodsForm(
  props: InformMethodsFormProps
): React.ReactElement {
  const [informMethodList, setInformMethodList] = useState<InformMethod[]>([]);

  useEffect(() => {
    const fetchInformMethodList = async (): Promise<void> => {
      const informMethodList = await CustomSenderApi_listSupportInform();
      setInformMethodList(informMethodList.data);
    };
    fetchInformMethodList();
  }, []);

  return (
    <FormItemWrapper {...props}>
      <InformMethodsFormItem informMethodList={informMethodList} />
    </FormItemWrapper>
  );
}
