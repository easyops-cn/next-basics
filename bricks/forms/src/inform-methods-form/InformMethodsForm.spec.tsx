import React from "react";
import { render } from "@testing-library/react";

import { InformMethodsFormItem } from "./InformMethodsForm";

import { CustomSenderApi_listSupportInform } from "@next-sdk/msgsender-sdk";

import { InformMethod } from "../interfaces";

jest.mock("@next-sdk/msgsender-sdk");

describe("InformMethodsForm", () => {
  beforeEach(() => {
    const informMethodList: InformMethod[] = [
      {
        inform_type: "phone",
        enable: false,
        col_of_cmdb_user_object: "user_tel",
        description: "电话",
        plugin_name: "phone",
      },
      {
        inform_type: "message",
        enable: false,
        col_of_cmdb_user_object: "user_tel",
        description: "短信",
        plugin_name: "message",
      },
      {
        inform_type: "email",
        enable: true,
        col_of_cmdb_user_object: "user_email",
        description: "邮件",
        plugin_name: "mail",
      },
      {
        inform_type: "wework",
        enable: true,
        col_of_cmdb_user_object: "wework_userid",
        description: "企业微信",
        plugin_name: "wework",
      },
    ];
    (CustomSenderApi_listSupportInform as jest.Mock).mockResolvedValue({
      data: informMethodList,
    });
  });

  it("should work", async () => {
    const result = render(<InformMethodsFormItem />);
    const asFragment = result.asFragment;

    await (global as any).flushPromises();
    expect(asFragment()).toMatchSnapshot();
  });
});
