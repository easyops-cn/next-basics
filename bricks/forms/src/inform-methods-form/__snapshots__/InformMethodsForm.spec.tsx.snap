// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InformMethodsForm should work 1`] = `
<DocumentFragment>
  <span />
</DocumentFragment>
`;
